Feat/unittests

SGSSGene opened the pull request:
- convert Param_test.cpp unittest to gtest - most things work
  - everything exception based, is currently not working and not tested
  - the logging framework is only integrated by redirecting std::cerr and std::cout

smehringer commented:
can be rebased now

smehringer reviewed (changes requested):
First round.

Please replace all `EXPECT_EQ(..., true)` with `EXPECT_TRUE(...)`
Then I'll review the rest.

SGSSGene commented on test/ParamValue_test.cpp:
```diff
@@ -5,829 +5,695 @@
+TEST(ParamValue, ctor_const_vector_string) {
+  std::vector<std::string> sl = {"test string", "test string 2"};
+  tdl::ParamValue d(sl);
+  EXPECT_EQ(d == sl, true);
```
Ah, yes, this is an artifact on how the unittest was written before. I can change this.

SGSSGene commented on test/ParamValue_test.cpp:
```diff
@@ -5,829 +5,695 @@
+TEST(ParamValue, ctor_long_double) {
+  long double x = -3.4L;
+  tdl::ParamValue d(x);
+  // Note: The implementation uses typedef double (as opposed to float, double, long double.)
+  EXPECT_FLOAT_EQ((double)d, -3.4L);
```
I am using c-casts because the unittests in openms where using c-casts and I didn't wanted to break anything by changing semantics.

smehringer replied:
ah ok. fine then!

SGSSGene commented on test/ParamValue_test.cpp:
```diff
@@ -5,829 +5,695 @@
+TEST(ParamValue, ctor_short_int) {
+  short int n = -3000;
+  tdl::ParamValue d(n);
+  EXPECT_EQ((short int)d, -3000);
```
What do you mean by "use the variable itself for testing"?

smehringer replied:
`n` instead of `-3000` in this case

SGSSGene commented on test/ParamValue_test.cpp:
```diff
@@ -5,829 +5,695 @@
+  EXPECT_EQ((std::string)d, "test char");
+}
+
+TEST(ParamValue, ctor_const_std_string) {
+  std::string s = "test string";
+  tdl::ParamValue d(s);
+  EXPECT_EQ(d, "test string");
+}
```
Ah, yes of course, the older version had a string "ParamValue(const std::string)" and I converted this to "ctor_const_std_string". That is confusing. I will adjust that.

SGSSGene commented:
I know, the unit tests are not perfect yet. Many quirks come from porting the test to gtest but keeping the code as close as possible to the original semantic and structure.

- I replaced most EXPECT_EQ(*, true) with EXPCT_TRUE(*) (same for EXPECT_FALSE)
- I used EXPECT_DOUBLE_EQ where doubles are involved
- changed all places I found to static_cast instead of c-style casts
- made a few unit tests name clearer by removing the "_const_" attribute.
